### PR TITLE
feat(linter): no func assign

### PR DESCRIPTION
**Pull request by RiESAEX**
implement [no-func-assign](https://eslint.org/docs/latest/rules/no-func-assign)
#123

### CODE REVIEW

File: crates/oxc_semantic/src/binder.rs [hunk 1]
```diff
@@ -96,7 +96,9 @@ impl<'a> Binder for Function<'a> {
             if !scope.strict_mode && matches!(builder.parent_kind(), AstKind::IfStatement(_)) {
                 // Do not declare in if single statements,
                 // if (false) function f() {} else function g() { }
-            } else if self.r#type == FunctionType::FunctionDeclaration {
+            } else if self.r#type == FunctionType::FunctionDeclaration
+                || self.r#type == FunctionType::FunctionExpression
```
Comment (RiESAEX) [inline crates/oxc_semantic/src/binder.rs:100]: this is for 
```js
var a = function foo() { foo = 123; };
```
But I'm not sure that's the right change.

Comment (Boshen) [reply crates/oxc_semantic/src/binder.rs:100]: Can you run `cargo coverage` and see if it breaks anything

Comment (Boshen) [reply crates/oxc_semantic/src/binder.rs:100]: You can also comment out this test case if it's not that easy to implement.

**Review by Boshen — APPROVED** [review]

**Comment by Boshen** [comment]
Thank you very much!

**Pull request merged**